### PR TITLE
Add LFD harnesses for roadmap blue lanes

### PR DESCRIPTION
**Pull request by serrrfirat**
## Summary

This PR adds the first reusable Loss Function Development (LFD) infrastructure for the IronClaw roadmap blue lanes.

It is intentionally not a feature implementation PR. It is the evaluation, scoring, and launch infrastructure that lets us hand individual roadmap lanes to implementation agents without asking those agents to invent their own tests, loss functions, scorers, or answer keys.

The PR adds:

- a shared LFD schema for cases, runner outcomes, and sealed contracts
- shared scorer/lint/probe/status scripts under `lfd/_shared/`
- a data-driven Rust integration runner under `tests/integration/lfd/`
- profile skeletons for the first Wave-1 lanes
- designer-authored LFD packages for the Wave-1 roadmap lanes
- a smoke pilot proving the runner/scorer/probe loop works end-to-end
- roadmap documentation that explains assignment order, ownership boundaries, and launch checklists

## Why We Are Doing This

The blue-lane roadmap work is too large and too easy to self-grade incorrectly if each implementation agent creates its own eval and success criteria.

The original lane documents were useful as LFD goals, but they still had a major launch-blocking problem: the implementation agent for a lane could also materialize the harness, eval data, and answers. That creates a self-graded optimization target. A literal optimizer can satisfy that target by weakening the scorer, memorizing visible evals, overfitting leaked feedback, or designing the test around its own implementation instead of the product behavior we actually want.

This PR separates roles:

- **Designer-owned artifacts:** eval cases, sealed contracts, caps, pins, scorer behavior, holdout placement, and launch checklist.
- **Implementation-agent artifacts:** product code, tests, the lane's `LOG.md`, and exactly one runner profile file for that lane.

That gives us a safer handoff model. Agents can optimize against the lane, but they do not own the judge.

The goal is to make genuine product capability the cheapest path to a high score.

## What This Is Used For

This infra is meant to be the handoff layer between the roadmap and implementation agents.

For each feature lane, the repo now has an `lfd/<lane>/` package containing:

- `goal.md` — the optimization goal and constraints
- `spec.md` — the inner-loop build spec
- `eval/dev/cases/*.json` — visible eval inputs
- `harness/answers.dev.json` — sealed dev contracts, canary-tokened and read-banned for optimizers
- `harness/caps.json` — capacity caps and anti-cheat constraints
- `harness/pins.json` — hash pins for scorer/runner/support sources
- `harness/score.sh` — scores runner outcomes
- `harness/lint.sh` — enforces canaries, caps, pins, and answer-literal overlap checks
- `harness/probe.sh` — generates deterministic perturbations to detect memorization
- `harness/status.sh` — reports elapsed time, score history, holdout budget, spend, and trend
- `LOG.md` — optimizer cycle log

For each lane, the Rust runner resolves a profile under:

```text
tests/integration/lfd/profiles/<lane>.rs
```

That profile is the adapter between generic LFD cases and the real IronClaw integration harness. The profile is the only runner file an implementation agent is expected to modify for its lane.

The expected implementation loop is:

1. Pick one lane.
2. Read `lfd/<lane>/goal.md` and `lfd/<lane>/spec.md`.
3. Implement Stage 0 until the normal code/test gate is green.
4. Wire `tests/integration/lfd/profiles/<lane>.rs` so cases return `status: "ran"` instead of `unsupported`.
5. Run the LFD runner to produce outcomes.
6. Run `harness/score.sh` on those outcomes.
7. Run `harness/probe.sh` to check whether the change generalized or memorized the dev set.
8. Log the hypothesis/result in `LOG.md`.
9. Commit each optimization cycle.
10. Use holdout only for acceptance, aggregate-only and rate-limited.

### Shared Schema

`lfd/_shared/SCHEMA.md` defines three JSON contracts:

1. **Case** — visible eval input, including setup, scripted model turns, inbound messages, and post-run state queries.
2. **Outcome** — runner output, including status, replies, tool invocations, egress, events, gates, state query results, leak scan results, and runner hash.
3. **Contract** — sealed answer key, with required matchers, forbidden matchers, and ordered sequence checks.

The scoring formula is intentionally two-sided:

- missing required behavior reduces the numerator
- forbidden behavior applies penalties
- ordered-sequence violations zero the case
- `unsupported` scores zero
- hard errors score zero where configured

This prevents one-sided cheats like returning everything, returning nothing, or denying every request.

### Shared Scorer

`lfd/_shared/scorer/score_core.py` handles:

- dev scoring
- holdout scoring
- probe scoring
- capped dev feedback
- holdout rate limiting
- audit logging
- runner hash checks for holdout

Dev feedback is intentionally low resolution: aggregate score plus at most five worst case IDs. It does not reveal matcher failures or expected values.

Holdout feedback is stricter: aggregate-only, max three calls per 24h, audit-logged under the off-repo state root.

### Shared Lint

`lfd/_shared/scorer/lint_core.py` enforces:

- dev answer canaries do not appear outside the sealed answer file
- holdout canaries do not appear in repo-visible files
- lane-specific capacity caps
- pins against scorer/runner/support sources
- answer-literal overlap checks against newly added product/test code

Lint violations print only:

```text
VOID: constraint violation
```

Detailed findings go to the state root outside the optimizer surface, so the lint does not become a membership oracle for mining answer values.

### Shared Probe

`lfd/_shared/scorer/probe_core.py` generates deterministic perturbations of visible dev inputs. This gives us a memorization gauge: if dev score improves but probe score collapses, the implementation is probably overfitting eval-shaped literals.

Probe output is generated on demand and intentionally not committed.

### Shared Status

`lfd/_shared/scorer/status_core.py` reports:

- elapsed time after a real `RUN START`
- dev score history
- holdout budget use
- spend ledger status
- score trend

Generated lane packages keep `RUN START` unset until an actual implementation loop begins.

### Rust Integration Runner

`tests/integration/lfd/` adds `reborn_lfd_runner`, a data-driven batch runner over the Reborn integration harness.

The runner:

- reads `LFD_CASES`
- executes each `*.json` case through the selected profile
- isolates case failures
- emits one outcome JSON per case under `LFD_OUT`
- reports `unsupported` rather than silently skipping missing profile behavior
- computes a runner hash from pinned scorer/runner/support sources

The runner is intentionally generic. Product-specific behavior is isolated in profiles.

### Pins

Each lane has `harness/pins.json` with hashes for the shared scorer, LFD runner, and integration support files.

Dev scoring records pin mismatches. Holdout scoring treats pin mismatch as a hard violation. This prevents an optimizer from editing the judge or extraction boundary and then claiming a holdout score.

## What This PR Adds By Lane

The first Wave-1 package set is now present.

Full packages:

| Lane | Dev cases | Dev contracts | Holdout cases | Holdout contracts |
| --- | ---: | ---: | ---: | ---: |
| cleanup-old-architecture | 30 | 30 | 3 | 3 |
| custom-build-tools | 30 | 30 | 12 | 12 |
| missions | 30 | 30 | 12 | 12 |
| memory-placement-product-layer | 30 | 30 | 12 | 12 |

Complete pilot slices:

| Lane | Dev cases | Dev contracts | Holdout cases | Holdout contracts |
| --- | ---: | ---: | ---: | ---: |
| slack-channel | 10 | 10 | 3 | 3 |
| secrets-skills-tools | 10 | 10 | 3 | 3 |
| user-voice-model | 10 | 10 | 3 | 3 |

Smoke pilot:

| Package | Purpose |
| --- | --- |
| `lfd/smoke-pilot` | Small proof that the runner, scorer, lint, probe, pins, and status scripts work end-to-end before launching real lanes. |

## Holdout Handling

Holdout cases and answers are intentionally not committed to this PR.

They were generated locally under:

```text
/Volumes/NVME/ironclaw-lfd/holdout/<lane>/
```

That is by design. Holdout answers should live outside the optimizer-readable repo. The scorer expects them under the configured state root when running holdout mode.

Before a real lane launch in a different environment, the designer/operator needs to provision the lane's holdout bundle into that environment's `LFD_STATE_ROOT`.

## What This Is Not

This PR does not implement the roadmap features themselves.

Most lane profiles are skeletons that compile and intentionally return `unsupported`. That is expected. It lets the package exist and score as zero until the lane implementation wires the profile to real product behavior.

This PR is also not a replacement for specialized benchmark harnesses such as the latency evidence harness in PR #5727. The LFD infra is the reusable behavior-contract harness for IronClaw feature lanes. Performance- or latency-specific work can still use specialized harnesses, preferably following the same goal/spec/score/probe/status conventions.

## How To Use This After Merge

Example for the smoke pilot:

```sh
CARGO_TARGET_DIR=target-lfd \
LFD_CASES=lfd/smoke-pilot/eval/dev/cases \
LFD_OUT=/tmp/lfd-smoke-out \
cargo test --test reborn_lfd_runner -- --nocapture

lfd/smoke-pilot/harness/score.sh \
  --repo-root . \
  --state-root /tmp/lfd-smoke-state \
  --outcomes /tmp/lfd-smoke-out

lfd/smoke-pilot/harness/probe.sh
```

Example lane launch shape:

```sh
# 1. Implement / wire the lane profile first.
$EDITOR tests/integration/lfd/profiles/<lane>.rs

# 2. Run the profile through the generic runner.
CARGO_TARGET_DIR=target-lfd \
LFD_CASES=lfd/<lane>/eval/dev/cases \
LFD_OUT=/tmp/lfd-<lane>-out \
cargo test --test reborn_lfd_runner -- --nocapture

# 3. Score the emitted outcomes.
lfd/<lane>/harness/score.sh \
  --repo-root . \
  --state-root /tmp/lfd-<lane>-state \
  --outcomes /tmp/lfd-<lane>-out

# 4. Generate perturbations and compare probe behavior.
lfd/<lane>/harness/probe.sh
```

## Validation Performed

The smoke pilot was deliberately run before broad lane package generation, then rerun after pins and runner-hash behavior were finalized.

Validation commands run locally:

```sh
python3 lfd/_shared/scorer/selftest/harness.py
```

Result:

```text
SELF-TEST: 12/12 PASS
```

```sh
cargo fmt --check
```

Result: passed.

```sh
CARGO_TARGET_DIR=target-lfd cargo test --test reborn_lfd_runner --no-run
```

Result: passed; runner compiled.

Smoke dev run:

```sh
CARGO_TARGET_DIR=target-lfd \
LFD_CASES=lfd/smoke-pilot/eval/dev/cases \
LFD_OUT=/tmp/lfd-smoke-out-current \
cargo test --test reborn_lfd_runner -- --nocapture

lfd/smoke-pilot/harness/score.sh \
  --repo-root . \
  --state-root /tmp/lfd-smoke-state-current \
  --outcomes /tmp/lfd-smoke-out-current
```

Result:

```text
2 smoke outcomes emitted with status: ran
score: 1.0000
```

Smoke probe run:

```sh
lfd/smoke-pilot/harness/probe.sh

CARGO_TARGET_DIR=target-lfd \
LFD_CASES=lfd/smoke-pilot/eval/probe/cases \
LFD_OUT=/tmp/lfd-smoke-probe-out-current \
cargo test --test reborn_lfd_runner -- --nocapture

lfd/smoke-pilot/harness/score.sh \
  --repo-root . \
  --state-root /tmp/lfd-smoke-state-current \
  --outcomes /tmp/lfd-smoke-probe-out-current \
  --probe lfd/smoke-pilot/eval/probe/map.json
```

Result:

```text
2 probe outcomes emitted with status: ran
score: 1.0000
gap_vs_dev: +0.0000
```

All lane wrappers were also checked after pins were filled:

```sh
lfd/<lane>/harness/lint.sh --repo-root . --state-root /tmp/lfd-<lane>-state
lfd/<lane>/harness/status.sh --state-root /tmp/lfd-<lane>-state
```

Result: all checked lane lint/status wrappers passed.

## Review Notes / Risks

This PR is large because it includes generated designer-owned eval packages and sealed dev contracts. The most important review surfaces are not every individual JSON line; they are:

1. Is the shared schema expressive enough for IronClaw feature lanes?
2. Does the scorer price both failure directions?
3. Does lint avoid leaking the exact constraint violation to the optimizer?
4. Is the runner/profile trust boundary understandable and hard to misuse?
5. Do lane packages keep implementation agents out of `harness/`, `eval/`, answers, caps, and pins?
6. Are the first Wave-1 packages good enough to start implementation agents, with smaller slices expanded before long-running optimization?

Known caveats:

- The runner compile is currently heavy because it uses the full integration target.
- The skeleton lane profiles return `unsupported` until implementation work wires them to production behavior.
- Slack, secrets, and voice are complete pilot slices, not full 30+ case packages yet.
- Holdout bundles are off-repo and must be provisioned to `LFD_STATE_ROOT` for real holdout scoring.
- This infra is strongest for structured behavior contracts: replies, tools, egress, events, gates, state queries, and leak scans. Pure UI pixel work or performance-only work may still need specialized harnesses.

## Next Steps After This PR

1. Review and merge this infrastructure PR.

2. Pick the first real lane to launch. Best first candidates are the full packages:
   - `custom-build-tools`
   - `missions`
   - `memory-placement-product-layer`
   - `cleanup-old-architecture`

3. For the selected lane, create a focused implementation branch where the agent is allowed to edit:
   - product/runtime code required by the lane
   - relevant tests
   - `lfd/<lane>/LOG.md`
   - exactly one runner profile file under `tests/integration/lfd/profiles/`

4. The first implementation milestone for any lane is not “high score.” It is:
   - Stage 0 tests green
   - the lane profile executes dev cases with `status: "ran"`
   - the scorer returns a real numeric baseline instead of all `unsupported`

5. Expand the smaller pilot slices before serious long-running optimization:
   - Slack: grow beyond 10 dev / 3 holdout
   - Secrets: grow beyond 10 dev / 3 holdout
   - Voice: grow beyond 10 dev / 3 holdout

6. Add lane-specific known-good / known-bad calibration outcomes for each package. The shared scorer self-test passes, but each lane should still prove its own contracts separate a valid outcome from a shortcut.

7. Add CI or a lightweight check that runs:
   - shared scorer self-test
   - `cargo test --test reborn_lfd_runner --no-run`
   - per-lane JSON/schema validation
   - per-lane lint/status wrapper smoke checks

8. After the first lane implementation is complete, run holdout from a clean checkout with a provisioned `LFD_STATE_ROOT`, then record only the aggregate holdout score in the lane log.

### CODE REVIEW

**Comment by serrrfirat** [comment]
Closing this draft for now. We need to clean up and reshape the LFD infra before reopening.
